src/shared/condition.c

```diff
@@ -353,6 +353,61 @@ static int condition_test_osrelease(Condition *c, char **env) {
         return true;
 }
 
+static int condition_test_dmi(Condition *c, char **env) {
+        int r;
+
+        assert(c);
+        assert(c->type == CONDITION_DMI);
+
+        for (const char *parameter = ASSERT_PTR(c->parameter);;) {
+                _cleanup_free_ char *key = NULL, *condition = NULL, *actual_value = NULL;
+                OrderOperator order;
+                const char *word;
+                bool matches;
+
+                r = extract_first_word(&parameter, &condition, NULL, EXTRACT_UNQUOTE);
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to parse parameter: %m");
+                if (r == 0)
+                        break;
+
+                /* parse_order() needs the string to start with the comparators */
+                word = condition;
+                r = extract_first_word(&word, &key, "!<=>", EXTRACT_RETAIN_SEPARATORS);
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to parse parameter: %m");
+                /* Ensure env-var-like key names */
+                if (r == 0 || isempty(word))
+                        return log_debug_errno(SYNTHETIC_ERRNO(EINVAL),
+                                        "Failed to parse parameter, key/value format expected: %m");
+
+                /* Do not allow whitespace after the separator */
+                order = parse_order(&word);
+                if (order < 0 || isempty(word) || strchr(WHITESPACE, *word) != NULL)
+                        return log_debug_errno(SYNTHETIC_ERRNO(EINVAL),
+                                        "Failed to parse parameter, key/value format expected: %m");
+
+
+                const char *p = strjoina("/sys/class/dmi/id/", ascii_strlower(key));
+                r = read_one_line_file(p, &actual_value);
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to read %s, ignoring: %m", p);
+
+                /* Might not be comparing versions, so do exact string matching */
+                if (order == ORDER_EQUAL)
+                        matches = streq_ptr(actual_value, word);
+                else if (order == ORDER_UNEQUAL)
+                        matches = !streq_ptr(actual_value, word);
+                else
+                        matches = test_order(strverscmp_improved(actual_value, word), order);
```

Comment:
dmi fields are typically not version strings, hence, does this really make sense to use strverscmp_improved()?

Reply:
I had the `xxx_version` fields in mind (e.g `bios_version`, `board_version`, `chassis_version`, `product_version`). Do you think this makes sense?

Reply:
> I had the xxx_version fields in mind (e.g bios_version, board_version, chassis_version, product_version). Do you think this makes sense?

I have my doubts about that. They are unlikely to follow the versioning syntax we typically use in Linux packaging, hence I would really not go as far as assuming we could clearly define an ordering for them generically.

The DMI matching code we have so far (in udev) goes via fnmatch() too pretty much. Hence I'd really stick to that, and not do version compares here. I think it's quite a stretch to assume version strings read from DMI are sanely comparable...

Reply:
I think this can be closed